### PR TITLE
Show warning when using the docker update mechanism

### PR DESCRIPTION
**Pull request by mynameisbogdan**
#### Database Migration
NO

#### Description
image

Also
- Prevent NullRef in IsPathValid for null paths
- Replace legacy master branch with main

### CODE REVIEW

**Comment by stevietv** [comment]
does this actually work in lidarr? 

image

**Comment by markus101** [comment]
Do docker images have the ability to tell Sonarr which branch to display updates for? If not then that should still be configurable.

**Comment by thezoggy** [comment]
just fyi in sab we actually had to change from using this docker detection because it doesnt work for everything anymore

now we just look to see if `.dockerenv` exists instead. which works for older+newer docker (Ubuntu 18.04 and 22.04).

**Comment by bakerboy448** [comment]
agree with taloth; the detect docker fix is good.  but docker / other third party maintainers maintainers should be using package_info appropriately to disable it and tell users how to update (with that said - some additional documentation on package_info would likely be useful)

**Comment by bakerboy448** [comment]
Run a script to run compose update or any various scripts that would update the container(s)

**Comment by Taloth** [comment]
> > But... disabling of the auto updater goes via package_info. It doesn't need to be hardcoded and users should be able to select Script option if they want to.
> 
> Also it seems like if it's docker it doesn't allow any type of updates, so this PR only updates the UI to reflect the current state.
> 
> https://github.com/Sonarr/Sonarr/blob/a645681cc8a3683876270af1d1a7b052a86e87e1/src/NzbDrone.Core/Update/InstallUpdateService.cs#L240-L244

That should not have been added imo. coz a container dev can specify the desired behavior in package_info. There's no reason to block it just because it's docker. And thus I'd rather revert that, even though it was added a year ago, possible as pull from Radarr. I'll talk to markus about it.

I agree package_info should be documented better, but the two major image maintainers know and use it.

**Comment by markus101** [comment]
I agree with Taloth here, seems like we shouldn't have added the disabling of updating for docker completely in favour of relying on `package_info` instead.

File: frontend/src/Settings/General/UpdateSettings.js [hunk 1]
```diff
@@ -83,8 +83,10 @@ function UpdateSettings(props) {
                 type={inputTypes.CHECK}
                 name="updateAutomatically"
                 helpText={translate('UpdateAutomaticallyHelpText')}
+                helpTextWarning={updateMechanism.value === 'docker' ? translate('AutomaticUpdatesAreDisabledWhenUsingTheDockerUpdateMechanism') : undefined}
```
Comment (markus101) [inline frontend/src/Settings/General/UpdateSettings.js:86]: This translation key is too verbose, something like `AutomaticUpdatedDisabledDocker` would be more appropriate, I've already discussed downstream syncing up with Q.

File: src/NzbDrone.Core/Localization/Core/en.json [hunk 2]
```diff
@@ -83,6 +83,7 @@
   "Automatic": "Automatic",
   "AutomaticAdd": "Automatic Add",
   "AutomaticSearch": "Automatic Search",
+  "AutomaticUpdatesAreDisabledWhenUsingTheDockerUpdateMechanism": "Automatic updates are disabled when using the Docker update mechanism. Update the container image instead. Contact your container maintainer for additional details.",
```
Comment (markus101) [inline src/NzbDrone.Core/Localization/Core/en.json:86]: > Contact your container maintainer for additional details.

What's the value in this? Seems like unnecessary noise.

**Review by markus101 — CHANGES_REQUESTED** [review]

File: frontend/src/Settings/General/UpdateSettings.js (hunk 1 above)
Comment (markus101) [reply frontend/src/Settings/General/UpdateSettings.js:86]: We don't want to use the English Translation as a key, so we shouldn't convert the English translation to a key.

File: src/NzbDrone.Core/Localization/Core/en.json (hunk 2 above)
Comment (bakerboy448) [reply src/NzbDrone.Core/Localization/Core/en.json:86]: userproof else they'd be coming to the sonarr discord?

Comment (markus101) [reply src/NzbDrone.Core/Localization/Core/en.json:86]: Potentially, but now we're sending them to the maintainer's discord when it's not really something they need to deal with either. We already have docker disabled for updates and don't have many people asking how to update it.

```suggestion
  "AutomaticUpdatesAreDisabledWhenUsingTheDockerUpdateMechanism": "Automatic updates are disabled when using the Docker update mechanism. You will need to update the container image outside of Sonarr.",
```

File: frontend/src/Settings/General/UpdateSettings.js [hunk 3]
```diff
@@ -83,8 +83,10 @@ function UpdateSettings(props) {
...
                 onChange={onInputChange}
                 {...updateAutomatically}
+                isDisabled={updateMechanism.value === 'docker'}
```
[the remark this answers is not in the document]
Comment (markus101) [reply frontend/src/Settings/General/UpdateSettings.js:89]: I think the latter

File: src/NzbDrone.Core/Localization/Core/en.json [hunk 4]
```diff
@@ -83,6 +83,7 @@
   "Automatic": "Automatic",
   "AutomaticAdd": "Automatic Add",
   "AutomaticSearch": "Automatic Search",
+  "AutomaticUpdatesDisabledDocker": "Automatic updates are disabled when using the Docker update mechanism. You will need to update the container image outside of Sonarr.",
```
Comment (markus101) [inline src/NzbDrone.Core/Localization/Core/en.json:86]: In an effort to make this re-usable:
```suggestion
  "AutomaticUpdatesDisabledDocker": "Automatic updates are disabled when using the Docker update mechanism. You will need to update the container image outside of {appName}.",
```

Then pass `{ appName: 'Sonarr' }` into `translate`

**Review by markus101 — CHANGES_REQUESTED** [review]

**Review by markus101 — APPROVED** [review]

File: frontend/src/Settings/General/UpdateSettings.js [hunk 5]
```diff
@@ -83,8 +85,10 @@ function UpdateSettings(props) {
                 type={inputTypes.CHECK}
                 name="updateAutomatically"
                 helpText={translate('UpdateAutomaticallyHelpText')}
+                helpTextWarning={updateMechanism.value === 'docker' ? translate('AutomaticUpdatesDisabledDocker', { appName: 'Sonarr' }) : undefined}
                 onChange={onInputChange}
                 {...updateAutomatically}
+                isDisabled={!canUpdateAutomatically}
```
Comment (markus101) [inline frontend/src/Settings/General/UpdateSettings.js:91]: Actually, as mentioned by @bakerboy448 script updating is technically possible, so this is wholly accurate and would disable that ability.

File: src/NzbDrone.Core/Localization/Core/en.json [hunk 6]
```diff
@@ -95,6 +95,7 @@
   "Automatic": "Automatic",
   "AutomaticAdd": "Automatic Add",
   "AutomaticSearch": "Automatic Search",
+  "AutomaticUpdatesDisabledDocker": "Automatic updates are disabled when using the Docker update mechanism. You will need to update the container image outside of {appName}.",
```
Comment (markus101) [inline src/NzbDrone.Core/Localization/Core/en.json:98]:
```suggestion
  "AutomaticUpdatesDisabledDocker": "Automatic updates are not directly supported when using the Docker update mechanism. You will need to update the container image outside of {appName} or use a script",
```

**Review by markus101 — CHANGES_REQUESTED** [review]
On further review I think disabling automatic updates completely accurate as it could be possible with a script and this would gatekeep that ability.

File: frontend/src/Settings/General/UpdateSettings.js (hunk 5 above)
Comment (markus101) [reply frontend/src/Settings/General/UpdateSettings.js:91]: If the update mechanism is `docker`, the user either needs to update via docker (outside of Sonarr) or as baker mentioned before, they could use a script to update docker through compose or some other tooling.

Disabling this means they would be unable to use a script to do the updating automatically.

File: src/NzbDrone.Core/Update/ConfigureUpdateMechanism.cs [hunk 7]
```diff
@@ -65,6 +65,12 @@ private void ChangeUpdateMechanism(UpdateMechanism updateMechanism)
             {
                 [nameof(_configFileProvider.UpdateMechanism)] = updateMechanism
             };
+
+            if (updateMechanism >= UpdateMechanism.External)
+            {
+                config.Add(nameof(_configFileProvider.UpdateAutomatically), false);
+            }
```
[the remark this answers is not in the document]
Comment (markus101) [reply src/NzbDrone.Core/Update/ConfigureUpdateMechanism.cs:72]: Yeah I think so, it'll keep turning it off if they've enabled it, but if we can do it when the value hasn't yet been written to the file then that would be fine.

File: frontend/src/Settings/General/UpdateSettings.js (hunk 5 above)
Comment (stevietv) [reply frontend/src/Settings/General/UpdateSettings.js]: > How can one update with docker compose from inside of a running container?

for example, on identifying an update you could call a script that triggers another container (watchtower) to update this one.

**Review by markus101 — APPROVED** [review]

**Pull request merged**